File: outro_projeto/pages/note-modal.page.ts
import { expect, type Locator, type Page } from '@playwright/test';

export type NoteInput = {
  title: string;
  description: string;
  category: 'Home' | 'Work' | 'Personal';
  completed: number;
};

export class NoteModalPage {
  readonly page: Page;
  readonly addModalTitle: Locator;
  readonly editModalTitle: Locator;
  readonly categorySelect: Locator;
  readonly completedCheckbox: Locator;
  readonly titleInput: Locator;
  readonly descriptionInput: Locator;
  readonly submitButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.addModalTitle = page.getByText('Add new note');
    this.editModalTitle = page.getByText('Edit note');
    this.categorySelect = page.getByTestId('note-category');
    this.completedCheckbox = page.getByTestId('note-completed');
    this.titleInput = page.getByTestId('note-title');
    this.descriptionInput = page.getByTestId('note-description');
    this.submitButton = page.getByTestId('note-submit');
  }

  async expectAddModalVisible(): Promise<void> {
    await expect(this.addModalTitle).toBeVisible();
    await expect(this.categorySelect).toBeVisible();
  }

  async expectEditModalVisible(): Promise<void> {
    await expect(this.editModalTitle).toBeVisible();
    await expect(this.categorySelect).toBeVisible();
  }

  async fillAndSubmit(note: NoteInput): Promise<boolean> {
    await this.categorySelect.selectOption(note.category);

    for (let i = 0; i < note.completed; i++) {
      await this.completedCheckbox.click();
    }

    await this.titleInput.fill(note.title);
    await this.descriptionInput.fill(note.description);

    const expectedCompleted = note.completed % 2 !== 0;
    await this.submitButton.click();
    return expectedCompleted;
  }

  async expectTitleValidationError(): Promise<void> {
    await expect(this.page.getByText('Title should be between 4 and 100 characters')).toBeVisible();
  }

  async expectDescriptionValidationError(): Promise<void> {
    await expect(this.page.getByText('Description should be between 4 and 1000 characters')).toBeVisible();
  }
}

File: tests/web/pages/note_modal_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class NoteModalPage:
    def __init__(self, driver: WebDriver):
        self.driver = driver
        self.wait = WebDriverWait(driver, 10)
        self.add_modal_title = (By.XPATH, "//h5[contains(text(), 'Add new note')]")
        self.edit_modal_title = (By.XPATH, "//h5[contains(text(), 'Edit note')]")
        self.category_select = (By.CSS_SELECTOR, '[data-testid="note-category"]')
        self.completed_checkbox = (By.CSS_SELECTOR, '[data-testid="note-completed"]')
        self.title_input = (By.CSS_SELECTOR, '[data-testid="note-title"]')
        self.description_input = (By.CSS_SELECTOR, '[data-testid="note-description"]')
        self.submit_button = (By.CSS_SELECTOR, '[data-testid="note-submit"]')

    def invalid_title_message_visible(self):
        try:
            error = self.driver.find_element(By.XPATH, "//*[contains(text(), 'Title should be between 4 and 100 characters')]")
            return error.is_displayed()
        except Exception:
            return False

    def invalid_description_message_visible(self):
        try:
            error = self.driver.find_element(By.XPATH, "//*[contains(text(), 'Description should be between 4 and 1000 characters')]")
            return error.is_displayed()
        except Exception:
            return False

    def expect_add_modal_visible(self):
        assert self.wait.until(EC.visibility_of_element_located(self.add_modal_title))
        assert self.wait.until(EC.visibility_of_element_located(self.category_select))

    def expect_edit_modal_visible(self):
        assert self.wait.until(EC.visibility_of_element_located(self.edit_modal_title))
        assert self.wait.until(EC.visibility_of_element_located(self.category_select))

    def fill_and_submit(self, note):
        self.wait.until(EC.visibility_of_element_located(self.category_select)).send_keys(note['category'])
        for _ in range(note.get('completed', 0)):
            self.wait.until(EC.element_to_be_clickable(self.completed_checkbox)).click()
        self.wait.until(EC.visibility_of_element_located(self.title_input)).clear()
        self.wait.until(EC.visibility_of_element_located(self.title_input)).send_keys(note['title'])
        self.wait.until(EC.visibility_of_element_located(self.description_input)).clear()
        self.wait.until(EC.visibility_of_element_located(self.description_input)).send_keys(note['description'])
        self.wait.until(EC.element_to_be_clickable(self.submit_button)).click()

    def expect_title_validation_error(self):
        error = self.driver.find_element(By.XPATH, "//*[contains(text(), 'Title should be between 4 and 100 characters')]")
        assert error.is_displayed()

    def expect_description_validation_error(self):
        error = self.driver.find_element(By.XPATH, "//*[contains(text(), 'Description should be between 4 and 1000 characters')]")
        assert error.is_displayed()
